pkg/planner/core/indexmerge_unfinished_path.go

```diff
@@ -0,0 +1,380 @@
+	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/planner/util"
+)
+
+// Note that at this moment, the implementation related to unfinishedAccessPath only aims to handle OR list nested in
+// AND list, which is like ... AND (... OR ... OR ...) AND ..., and build an MV IndexMerge access path. So some struct
+// definition and code logic are specially designed for this case or only consider this case.
+// This may be changed in the future.
+
+// unfinishedAccessPath is for collecting access filters for an access path.
+// It maintains the information during iterating all filters. Importantly, it maintains incomplete access filters, which
+// means they may not be able to build a valid range, but could build a valid range after collecting more access filters.
+// After iterating all filters, we can check and build it into a valid util.AccessPath.
+type unfinishedAccessPath struct {
+	index *model.IndexInfo
+
+	accessFilters []expression.Expression
+
+	// To avoid regression and keep the same behavior as the previous implementation, we collect access filters in two
+	// methods:
+	//
+	// 1. Use the same functions as the previous implementation to collect access filters. They are able to handle some
+	// complicated expressions, but the expressions must be able to build into a valid range at once (that's also what
+	// "finished" implies).
+	// In this case, idxColHasAccessFilter will be nil and initedAsFinished will be true.
+	//
+	// 2. Use the new logic, which is to collect access filters for each column respectively, gradually collect more
+	// access filters during iterating all filters and try to form a valid range at last.
+	// In this case, initedAsFinished will be false, and idxColHasAccessFilter will record if we have already collected
+	// a valid access filter for each column of the index.
+	idxColHasAccessFilter []bool
+	initedAsFinished      bool
+
+	// needKeepFilter means the OR list need to become a filter in the final Selection.
+	needKeepFilter bool
+
+	// Similar to AccessPath.PartialIndexPaths, each element in the slice is expected to build into a partial AccessPath.
+	// Currently, it can only mean an OR type IndexMerge.
+	indexMergeORPartialPaths []unfinishedAccessPathList
+}
+
+// unfinishedAccessPathList is for collecting access filters for a slice of candidate access paths.
+// This type is useful because usually we have several candidate index/table paths. When we are iterating the
+// expressions, we want to match them against all index/table paths and try to find access filter for every path. After
+// iterating all expressions, we check if any of them can form a valid range so that we can build a valid AccessPath.
+type unfinishedAccessPathList []*unfinishedAccessPath
+
+// generateUnfinishedIndexMergePathFromORList handle the nested OR list, and returns an unfinishedAccessPath, which
+// must be an index merge OR unfinished path.
+func generateUnfinishedIndexMergePathFromORList(
```

Comment:
Could you add a concrete example in the comments? For example, the input is `a>1 or a=1 or ...` and the output is `xxx`.